src/org/mozartoz/truffle/nodes/call/TailCallCatcherNode.java

```diff
@@ -0,0 +1,45 @@
+package org.mozartoz.truffle.nodes.call;
+
+import org.mozartoz.truffle.nodes.DerefNode;
+import org.mozartoz.truffle.nodes.OzNode;
+import org.mozartoz.truffle.runtime.OzArguments;
+import org.mozartoz.truffle.runtime.TailCallException;
+
+import com.oracle.truffle.api.dsl.CreateCast;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+
+public class TailCallCatcherNode extends OzNode {
+
+	@Child OzNode call;
+	@Child IndirectCallNode callNode = IndirectCallNode.create();
+
+	public TailCallCatcherNode(OzNode call) {
+		this.call = call;
+	}
+
+	@CreateCast("function")
+	protected OzNode derefFunction(OzNode var) {
+		return DerefNode.create(var);
+	}
+
+	@Override
+	public Object execute(VirtualFrame frame) {
+		TailCallException tailCall = null;
+		while (true) {
+			try {
+				if (tailCall == null) {
+					return call.execute(frame);
+				} else {
+					return callNode.call(frame,
+							tailCall.proc.callTarget,
+							OzArguments.pack(tailCall.proc.declarationFrame, tailCall.arguments)
+							);
```

Comment:
And then reuse the CallProcNode here, with something like
`executeCall(proc, arguments)`.
But CallProcNode needs to be adapted to allow taking direct arguments and not execute the arg nodes.